### ISSUE TITLE
Allow for adding additional functions to `_worker.js`

### ISSUE THREAD

**Issue by smitssjors** [sveltejs/kit#10496]
### Describe the problem

[Cloudflare Queues](https://developers.cloudflare.com/queues) is currently in open beta. It would be nice to be able to consume messages from the same worker that handles HTTP requests, as opposed to having to creating a separate worker. For a worker to consume from a queue a new `queue` function needs to be defined on the worker object.
```javascript
export default {
  async fetch(request, env, ctx) {
    // Handle request...
  },
  async queue(batch, env) {
    // Process batch...
  },
};
```
Currently there is no easy way to add this `queue` function to the woker object in the generated `_worker.js`.

Other Cloudflare services such as [Email Workers](https://developers.cloudflare.com/email-routing/email-workers/) follow a similar approach.

### Describe the proposed solution

I'd suggest a special `src/cloudflare-worker.js` file which would export functions such as `queue` which then during compilation get added to the `_worker.js` file.
```javascript
// src/cloudflare-worker.js

export const queue = (batch, env) =>  // Process batch...
```

### Importance

would make my life easier

**Comment by a-hariti** [comment]
How would the local DX look like in this context? @smitssjors 

I think CF pages works now as long as you don't use anything specific to CF,
miniflare still doesn't cover all the cases as well :(

**Comment by smitssjors** [comment]
@a-hariti It shouldn't have to be any different than it is [now](https://kit.svelte.dev/docs/adapter-cloudflare#bindings-testing-locally).

**Comment by a-hariti** [comment]
@smitssjors That's a terrible DX as it only works on the build output

you have to wait for a fresh build on each minute change

**Comment by smitssjors** [comment]
@a-hariti Yes I agree, but that is out of the scope of this issue. I think that fixing that would require substantial work on how adapters integrate with kit and vite. Currently they just adapt the output of building.

**Comment by ghostdevv** [comment]
It could be an interesting idea to come up with something similar to how adapter node works. We could have a `src/worker.ts` file that looks something like:

```js
import { handle } from 'some-virtual-module'

export default {
	fetch: handle
}
```

**Comment by ghostdevv** [comment]
Though I only think this would work when deploying to workers right?

**Comment by smitssjors** [comment]
@ghostdevv To me this looks like a nicer solution. The node adapter docs are not clear where the custom server will live e.g., does kit copy over all files in src or does a user need to copy them over manually?

> Though I only think this would work when deploying to workers right?

I think so.

**Comment by lsfratel** [comment]
+1

**Comment by victormihalache** [comment]
> It could be an interesting idea to come up with something similar to how adapter node works. We could have a `src/worker.ts` file that looks something like:
> 
> import { handle } from 'some-virtual-module'
> 
> export default {
> 	fetch: handle
> }

This is basically how it's done with other frameworks. See https://github.com/sveltejs/kit/issues/13692

**Comment by sergiocampama** [comment]
Does anyone interested in this have a working prototype for adding queues and schedule into the generated worker? Or is this feasible by forking the adapter to do what we need?

**Comment by teemingc** [comment]
This is coming in SvelteKit 3 soon

**Comment by sergiocampama** [comment]
Thanks @teemingc !

in the meanwhile I found a workaround for SvelteKit 2 if anyone is keen in testing it or improving it, it's in this gist: https://gist.github.com/sergiocampama/77b559523f403cbc700c71f330b3cccc

The basic approach is to wrap adapter-cloudflare and then modify the _worker.js output file by replacing it with a copy after adapter-cloudflare has run. Then creating fake requests and injecting the batch messages and cron values in the platform object, which can be checked by some middleware in hooks.server.ts

Ideally all of this should be replaced in SvelteKit 3 with a native solution, but this makes my life easier now :)

**Comment by alaixandro** [comment]
> This is coming in SvelteKit 3 soon

Thanks @teemingc for the update. Will v3 also allow auxiliaryworkers?

**Comment by teemingc** [comment]
> > This is coming in SvelteKit 3 soon
> 
> Thanks [@teemingc](https://github.com/teemingc) for the update. Will v3 also allow auxiliaryworkers?

It should. We're essentially using the Cloudflare Vite plugin and you'll be able to pass in options to do what the docs say https://developers.cloudflare.com/workers/vite-plugin/reference/programmatic-configuration/#auxiliary-workers

**Comment by alaixandro** [comment]
Would love to test when something is ready! Much appreciated @teemingc ✌️

**Comment by tskimmett** [comment]
Is there a rough timeline for sveltekit v3?

**Comment by teemingc** [comment]
We’re actively working on it

**Comment by davyzhang** [comment]
one way to work around this limitation is through service binding,

```json
			"services": [
				{
					"binding": "YOUR_SERVICE",
					"service": "dir-to-worker-code"
				}
                         ]

```
start these workers with wrangler dev
the main app's wrangler dev will pick them up showing `connected`

only http request, no tcp/socket, DO doesn't work

**Comment by tskimmett** [comment]
> one way to work around this limitation is through service binding,

Not sure about this. The main thing I and I think others in this thread are looking for is alternate entry points into the main sveltekit app, like cron and queues. A separate worker/code base doesn't really help with that.

**Comment by alaixandro** [comment]
Any work around until v3/v3-alpha is ready? No http requests, separate workers but same miniflare local env 

Like 

`configPath: "./wrangler.jsonc",
persistState: { path: "../../apps/web/.wrangler/state" },
auxiliaryWorkers: [{configPath:"../../apps/services/wrangler.jsonc" }],
`

I have very specific use case with sensitive data. 

Much appreciated for any feedback.
